spring-cloud-function-stream/src/main/java/org/springframework/cloud/function/stream/StreamListeningConsumerInvoker.java

```diff
@@ -16,64 +16,88 @@
 
-import java.util.function.Consumer;
 import java.util.function.Function;
 
-import reactor.core.publisher.Flux;
-
 import org.springframework.beans.factory.SmartInitializingSingleton;
 import org.springframework.cloud.function.context.FunctionInspector;
+import org.springframework.cloud.function.registry.FunctionCatalog;
 import org.springframework.cloud.stream.annotation.Input;
 import org.springframework.cloud.stream.annotation.StreamListener;
 import org.springframework.cloud.stream.converter.CompositeMessageConverterFactory;
-import org.springframework.cloud.stream.messaging.Processor;
+import org.springframework.cloud.stream.messaging.Sink;
 import org.springframework.messaging.Message;
 import org.springframework.messaging.converter.MessageConverter;
 
+import reactor.core.publisher.Flux;
+
 /**
  * @author Mark Fisher
  * @author Marius Bogoevici
  */
 public class StreamListeningConsumerInvoker implements SmartInitializingSingleton {
 
-	private final Consumer<Flux<?>> consumer;
-
-	private final String name;
-
 	private final FunctionInspector functionInspector;
 
 	private final CompositeMessageConverterFactory converterFactory;
 
 	private MessageConverter converter;
 
-	private Class<?> inputType;
+	private final FunctionCatalog functionCatalog;
+
+	private final String defaultEndpoint;
 
-	public StreamListeningConsumerInvoker(String name, Consumer<Flux<?>> consumer, FunctionInspector functionInspector,
-			CompositeMessageConverterFactory converterFactory) {
-		this.consumer = consumer;
-		this.name = name;
+	private final String[] names;
+
+	public StreamListeningConsumerInvoker(FunctionCatalog functionCatalog,
+			FunctionInspector functionInspector,
+			CompositeMessageConverterFactory converterFactory, String defaultEndpoint,
+			String... names) {
+		this.functionCatalog = functionCatalog;
 		this.functionInspector = functionInspector;
 		this.converterFactory = converterFactory;
+		this.defaultEndpoint = defaultEndpoint;
+		this.names = names;
 	}
 
 	@Override
 	public void afterSingletonsInstantiated() {
 		this.converter = this.converterFactory.getMessageConverterForAllRegistered();
-		this.inputType = this.functionInspector.getInputType(this.name);
 	}
 
 	@StreamListener
-	public void handle(@Input(Processor.INPUT) Flux<Message<?>> input) {
-		this.consumer.accept(input.map(convertInput()));
+	public void handle(@Input(Sink.INPUT) Flux<Message<?>> input) {
+		input.groupBy(this::select)
+				.filter(group -> functionCatalog.lookupConsumer(group.key()) != null)
+				.subscribe(group -> process(group.key(), group));
+	}
+
+	private void process(String name, Flux<Message<?>> flux) {
+		functionCatalog.lookupConsumer(name)
+				.accept(flux.map(message -> convertInput(name).apply(message)));
+	}
+
+	private String select(Message<?> input) {
+		String name = defaultEndpoint;
+		if (name == null) {
+			for (String candidate : names) {
+				Class<?> inputType = functionInspector.getInputType(candidate);
```

Comment:
Should we validate at context creation time that there are no collisions between supported input types across the Consumers/Functions. Otherwise, only the first candidate will ever "win" and it might not be obvious why certain functions are essentially ignored.